processing/src/main/java/org/apache/druid/segment/ReferenceCountedSegmentProvider.java

```
@@ -168,35 +139,63 @@ public short getAtomicUpdateGroupSize()
   }
 
   @Override
-  public Optional<Closeable> acquireReferences()
+  public Optional<Segment> acquireReference()
   {
-    return incrementReferenceAndDecrementOnceCloseable();
+    final Optional<Closeable> reference = incrementReferenceAndDecrementOnceCloseable();
+    return reference.map(ReferenceClosingSegment::new);
   }
 
-  @Override
-  public void validateOrElseThrow(PolicyEnforcer policyEnforcer)
+  private boolean includeRootPartitions(ReferenceCountedSegmentProvider other)
   {
-    policyEnforcer.validateOrElseThrow(this, null);
+    return startRootPartitionId <= other.startRootPartitionId && endRootPartitionId >= other.endRootPartitionId;
   }
 
-  @Override
-  public <T> T as(Class<T> clazz)
+  /**
+   * Wraps a {@link Segment} and closes reference to this segment, decrementing the counter instead of closing the
+   * segment itself
+   */
+  public final class ReferenceClosingSegment extends WrappedSegment
   {
-    if (isClosed()) {
-      return null;
+    private final Closeable referenceCloseable;
+    private boolean isClosed;
+
+    private ReferenceClosingSegment(Closeable referenceCloseable)
+    {
+      super(baseObject);
+      this.referenceCloseable = referenceCloseable;
     }
-    return baseObject.as(clazz);
-  }
 
-  @Override
-  public boolean isTombstone()
-  {
-    return baseObject.isTombstone();
-  }
+    @Nullable
+    @Override
+    public <T> T as(@Nonnull Class<T> clazz)
+    {
+      if (isClosed) {
```

Might be worth commenting that this class is not thread safe, and it's designed to run on a single thread. And maybe we could throw an exception here?

---

oops, I meant to add the thread-safety note to the javadocs of `ReferenceCountedObjectProvider`, that the provider must be thread-safe but the object returned by `acquireReference` has no such guarantee, will add there :+1:

---

re the exception, I think it would be ok, I was mostly thinking of being consistent with the old behavior of `ReferenceCountingSegment`, but changing it to be more opinionated sounds fine too